src/storage/ob_partition_parallel_merge_ctx.cpp

```diff
@@ -160,6 +162,473 @@ int ObParallelMergeCtx::init_serial_merge()
+      ret = OB_ERR_UNEXPECTED;
+      STORAGE_LOG(WARN, "Unexpected bad ranges.", K(ret));
+    } else {
+      STORAGE_LOG(INFO, "Succeeded to get ranges by the base sstable.", K(row_cnt), K(macro_block_cnt), K(ranges));
+    }
+  }
+
+  return ret;
+}
+
+// To get row-key ranges by iterating over rows
+static int get_ranges_by_iteration(
+    common::ObIArray<common::ObStoreRange>& ranges, ObIStoreRowIterator& iter,
+    common::ObIAllocator& allocator, const int64_t num_rows_per_range)
+{
+  int ret = OB_SUCCESS;
+
+  int64_t row_cnt = 0;
+  common::ObStoreRange range;
+  range.get_end_key().set_min();
+  range.set_left_open();
+  range.set_right_closed();
+  common::ObStoreRowkey rkey;
+  const ObStoreRow *row = NULL;
+  int count = 0;
+  int64_t t1 = ObTimeUtil::current_time();
+  while (OB_SUCC(ret) && OB_SUCC(iter.get_next_row(row))) {
+    count++;
+    if (count >= num_rows_per_range) {
+      row_cnt += count;
+      count = 0;
+      rkey.assign(row->row_val_.cells_, row->row_val_.count_);
+      if (
+          OB_FAIL(range.get_end_key().deep_copy(range.get_start_key(), allocator)) ||
+          OB_FAIL(rkey.deep_copy(range.get_end_key(), allocator))
+      ) {
+        STORAGE_LOG(WARN, "Failed to deep copy rowkey.", K(ret));
+      } else {
+        if (OB_FAIL(ranges.push_back(range))) {
+          STORAGE_LOG(WARN, "Failed to push range.", K(ret));
+        }
+      }
+    }
+  }
+  double time_elapsed = (ObTimeUtil::current_time() - t1) / 1000000.0;
+  row_cnt += count;
+  if (OB_ITER_END == ret) {
+    ret = OB_SUCCESS;
+  }
+  // if (OB_SUCC(ret)) {
+  //   STORAGE_LOG(WARN, "SCANTBL.", K(row_cnt), K(time_elapsed));
+  // }
+  // handle the last range
+  if (OB_SUCC(ret)) {
+    if (count > 0) {
+      if (OB_FAIL(range.get_end_key().deep_copy(range.get_start_key(), allocator))) {
+        STORAGE_LOG(WARN, "Failed to deep copy rowkey.", K(ret));
+      } else {
+        range.get_end_key().set_max();
+        range.set_right_open();
+        if (OB_FAIL(ranges.push_back(range))) {
+          STORAGE_LOG(WARN, "Failed to push range.", K(ret));
+        }
+      }
+    } else if (ranges.empty()) {
+      range.set_whole_range();
+      if (OB_FAIL(ranges.push_back(range))) {
+        STORAGE_LOG(WARN, "Failed to push range.", K(ret));
+      }
+    } else {
+      ranges.at(ranges.count() - 1).get_end_key().set_max();
+      ranges.at(ranges.count() - 1).set_right_open();
+    }
+  }
+
+  return ret;
+}
+
+int ObParallelMergeCtx::get_ranges_by_inc_data(
+    ObSSTableMergeCtx& merge_ctx, common::ObIArray<common::ObStoreRange>& ranges,
+    const common::ObStoreRowkey &start_key)
+{
+  int ret = OB_SUCCESS;
+
+  const uint64_t table_id = merge_ctx.table_schema_->get_table_id();
+  int64_t row_cnt = 0;
+  if (OB_UNLIKELY(merge_ctx.tables_handle_.get_count() <= 1)) {
+    // only have the base sstable, no inc data
+    row_cnt = 0;
+    STORAGE_LOG(INFO, "No inc sstable.", K(table_id), K(merge_ctx.tables_handle_.get_count()));
```

Comment:
parallel merge is also needed if the full merge is triggered even if there is no  inc tables